llvm/utils/TableGen/AIEVariableInstrItineraryEmitter.cpp

```diff
@@ -0,0 +1,176 @@
+
+  std::map<std::vector<std::pair<llvm::StringRef, llvm::StringRef>>,
+           std::vector<llvm::StringRef>>
+      UniqueRegItineraryPairs;
+  std::map<int, std::vector<llvm::StringRef>> UniqueOperandIdx;
+
+private:
+  RecordKeeper &Records;
+  CodeGenTarget Target;
+};
+
+} // namespace
+
+AIEVariableInstrItineraryEmitter::AIEVariableInstrItineraryEmitter(
+    RecordKeeper &R)
+    : Records(R), Target(R) {}
+
+void AIEVariableInstrItineraryEmitter::emitOperandIndexInfo(raw_ostream &OS) {
+  const std::string CurrentNamespace = Target.getName().str();
+  const std::string EndOfFunction("\n}\n\n");
+
+  // If there are no instructions with alternate itineraries, we do not need to
+  // dump getOperandIdx() function.
+  if (UniqueOperandIdx.empty())
+    return;
+
+  OS << "static std::optional<unsigned> ";
+  OS << "getOperandIdx(const unsigned Opcode) {\n";
+  OS << "  switch (Opcode) {\n";
+  OS << "  default:\n";
+  OS << "    return {};\n";
+  for (const auto &[OperandIdx, Instrs] : UniqueOperandIdx) {
+    for (const auto &Instr : Instrs) {
+      OS << "  case " << CurrentNamespace << "::" << Instr << ":\n";
+    }
+    OS << "    return " << OperandIdx << ";\n";
+  }
+  OS << "  }" << EndOfFunction;
+}
+
+void AIEVariableInstrItineraryEmitter::emitAltItineraryInfo(raw_ostream &OS) {
+  const std::string CurrentNamespace = Target.getName().str();
+  const std::string EndOfFunction("\n}\n\n");
+
+  OS << "unsigned " << CurrentNamespace << "InstrInfo::getSchedClass(\n";
+  OS << "    const MCInstrDesc &Desc,\n";
+  OS << "    iterator_range<const MachineOperand *> Operands) const {\n";
+  if (UniqueRegItineraryPairs.empty()) {
+    OS << "  return Desc.getSchedClass();" << EndOfFunction;
+    return;
+  }
+
+  OS << "  std::optional<unsigned> OperandIdx = "
+        "getOperandIdx(Desc.getOpcode());\n";
+  OS << "  if (!OperandIdx.has_value())\n";
+  OS << "    return Desc.getSchedClass();\n\n";
+
+  OS << "  const MachineOperand &MO = *std::next(Operands.begin(), "
+        "OperandIdx.value());\n";
+  OS << "  const Register Reg = MO.getReg();\n";
+  OS << "  assert(MO.isReg() && \"Operand must be of Reg type\");\n\n";
+
+  OS << "  const MachineInstr *MI = MO.getParent();\n";
+  OS << "  const MachineRegisterInfo &MRI =\n";
+  OS << "      "
+        "MI->getParent()->getParent()->getRegInfo();\n";
+  OS << "  const TargetRegisterClass *RC =\n";
+  OS << "      "
+        "Reg.isVirtual() ? MRI.getRegClass(Reg) : nullptr;\n\n";
+
+  OS << "  switch (Desc.getOpcode()) {\n";
+  OS << "  default:\n";
+  OS << "    return Desc.getSchedClass();\n";
+  for (const auto &[RegItineraryPairs, Instrs] : UniqueRegItineraryPairs) {
+    for (const auto &Instr : Instrs)
+      OS << "  case " << CurrentNamespace << "::" << Instr << ":\n";
+    OS << "  {\n";
+    for (const auto &[RegClass, Itinerary] : RegItineraryPairs) {
+      OS << "    if (regClassMatches(" << CurrentNamespace << "::" << RegClass
```

Comment:
As mentioned before, please use something like `AIEBaseRegisterInfo::regClassMatches` to make it obvious where that function comes from.